modules/generic-scala3/src/main/scala/pureconfig/generic/scala3/HintsAwareProductConfigWriterDerivation.scala

```diff
@@ -0,0 +1,73 @@
+package pureconfig
+package generic
+package scala3
+
+import scala.collection.JavaConverters.given
+import scala.compiletime._
+import scala.compiletime.ops.int._
+import scala.deriving.Mirror
+import scala.quoted._
+
+import com.typesafe.config.{ConfigValue, ConfigValueFactory}
+
+import pureconfig.generic.derivation.Utils
+
+trait HintsAwareProductConfigWriterDerivation { self: HintsAwareConfigWriterDerivation =>
+
+  inline def deriveProductWriter[A](using pm: Mirror.ProductOf[A], ph: ProductHint[A]): ConfigWriter[A] =
+    inline erasedValue[A] match {
+      case _: Tuple =>
+        new ConfigWriter[A] {
+          def to(a: A): ConfigValue = {
+            val values = writeTuple[pm.MirroredElemTypes, 0](a.asInstanceOf[Product])
+
+            ConfigValueFactory.fromIterable(values.asJava)
+          }
+        }
+
+      case _ =>
+        new ConfigWriter[A] {
+          def to(a: A): ConfigValue = {
+            val labels = Utils.transformedLabels(identity).toVector
+            val values = writeCaseClass[pm.MirroredElemTypes, 0, A](a.asInstanceOf[Product], labels)
+
+            ConfigValueFactory.fromMap(values.toMap.asJava)
+          }
+        }
+    }
+
+  private inline def writeTuple[T <: Tuple, N <: Int](product: Product): List[ConfigValue] =
+    inline erasedValue[T] match {
+      case _: (h *: t) =>
+        val n = constValue[N]
+        val value = product.productElement(n).asInstanceOf[h]
+        val head = summonConfigWriter[h].to(value)
+        val tail = writeTuple[t, N + 1](product)
+
+        head :: tail
+
+      case _: EmptyTuple => Nil
+    }
+
+  private inline def writeCaseClass[T <: Tuple, N <: Int, A: ProductHint](
+      product: Product,
+      labels: Vector[String]
+  ): List[(String, ConfigValue)] =
+    inline erasedValue[T] match {
+      case _: (h *: t) =>
+        val n = constValue[N]
+        def value[T] = product.productElement(n).asInstanceOf[T]
+
+        val valueOpt = summonConfigWriter[h] match {
+          case writer: WritesMissingKeys[h] => writer.toOpt(value)
```

Comment:
As written here the code is somewhat dangerous; `value[T]` can return any type whatsoever so it can be used anywhere, and if I'm reading this correctly it runs at runtime, not at compile time. Using `h` consistently across this function is probably safer.

```suggestion
        val value = product.productElement(n).asInstanceOf[h]

        val valueOpt = summonConfigWriter[h] match {
          case writer: WritesMissingKeys[`h` @unchecked] => writer.toOpt(value)
```

Reply:
Good point, fixed!